src/cli/Secrets.ts

```diff
@@ -128,37 +156,129 @@ function makeRemoveSecretCommand() {
+        const options = cmd.opts()
+
+        const client = PolykeyAgent.connectToAgent();
+        const nodePath = determineNodePath(options);
+
+        const isVerbose: boolean = options.verbose ?? false;
+        const command: string | undefined = options.command;
+        const run: string | undefined = options.run;
+
+        const secretPathList: string[] = Array.from<string>(cmd.args.values())
+
+
+        if (secretPathList.length < 1) {
+          throw Error("please specify at least one secret")
+        }
+
+        // Parse secret paths in list
+        const parsedPathList: { vaultName: string, secretName: string, variableName: string }[] = []
+        for (const path of secretPathList) {
+          if (!pathRegex.test(path)) {
+            throw Error(`secret path was not of the format '<vaultName>:<secretName>[=<variableName>]': ${path}`)
+          }
+          const [_, vaultName, secretName, variableName] = path.match(pathRegex)!
+          parsedPathList.push({
+            vaultName,
+            secretName,
+            variableName: variableName ?? secretName.toUpperCase().replace('-', '_')
+          })
+        }
+
+        const secretEnv = { ...process.env }
+        try {
+          // Get all the secrets
+          for (const obj of parsedPathList) {
+            const secret = await client.getSecret(nodePath, obj.vaultName, obj.secretName);
+            secretEnv[obj.variableName] = secret.toString()
+          }
+        } catch (err) {
+          throw Error(`Error when retrieving secret: ${err.message}`);
+        }
+        try {
+          const shellPath = process.env.SHELL!
+          const args: string[] = []
+          if (command && run) {
+            throw Error('only one of --command or --run can be specified')
+          } else if (command) {
+            args.push('-i')
+            args.push('-c')
+            args.push(`"${command}"`)
+          } else if (run) {
+            args.push('-c')
+            args.push(`"${run}"`)
+          }
+          const shell = spawn(shellPath, args, {
+            stdio: 'inherit',
+            env: secretEnv,
+            shell: true
+          })
+          shell.on('close', (code) => {
+            if (code != 0) {
+              pkLogger(`polykey: environment terminated with code: ${code}`, PKMessageType.WARNING)
+            }
+          })
+        } catch (err) {
+          throw Error(`Error when running environment: ${err.message}`);
+        }
+      }),
+    );
+}
+
 function makeSecretsCommand() {
   return new commander.Command('secrets')
     .description('manipulate secrets for a given vault')
     .addCommand(makeListSecretsCommand())
     .addCommand(makeNewSecretCommand())
     .addCommand(makeUpdateSecretCommand())
-    .addCommand(makeRemoveSecretCommand())
-    .addCommand(makeGetSecretCommand());
+    .addCommand(makeDeleteSecretCommand())
+    .addCommand(makeGetSecretCommand())
+    .addCommand(makeSecretEnvCommand());
```

Comment:
I think we should abstract alot of the CLI commands into separate files and modularize them. These files like `cli/Vaults` or `cli/Secrets` can act like top level sub commands with each handler being something lower level. I'm a big fan of MVC style, at least for the controllers themselves that can be instantiated with DI & IoC container. That way commands can share state too. And potentially we can compose commands later in the future.

The Haskell onion layering may be a good idea. https://www.parsonsmatt.org/2018/03/22/three_layer_haskell_cake.html and https://news.ycombinator.com/item?id=12604767

Note that these are architectural issues to be resolved later in a separate issue.

Reply:
https://github.com/MatrixAI/js-polykey/issues/83